File: mvp/lib-core-mvp/src/test/java/MockPresenterTest.java
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.powermock.api.mockito.PowerMockito;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import io.reactivex.Completable;
import io.reactivex.Maybe;
import io.reactivex.Observable;
import io.reactivex.Scheduler;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;
import ru.surfstudio.android.connection.ConnectionProvider;
import ru.surfstudio.android.core.mvp.error.ErrorHandler;
import ru.surfstudio.android.core.mvp.presenter.BasePresenter;
import ru.surfstudio.android.core.mvp.presenter.BasePresenterDependency;
import ru.surfstudio.android.core.ui.event.BaseScreenEventDelegateManager;
import ru.surfstudio.android.core.ui.navigation.activity.navigator.ActivityNavigator;
import ru.surfstudio.android.core.ui.state.ActivityScreenState;
import ru.surfstudio.android.logger.Logger;
import ru.surfstudio.android.rx.extension.ActionSafe;
import ru.surfstudio.android.rx.extension.ConsumerSafe;
import ru.surfstudio.android.rx.extension.scheduler.SchedulersProvider;

import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.verify;
import static org.mockito.internal.verification.VerificationModeFactory.only;
import static org.powermock.api.mockito.PowerMockito.mock;

/**
 * unit tests for {@link BasePresenter}
 **/

@RunWith(PowerMockRunner.class)
@PrepareForTest(Logger.class)
public class MockPresenterTest {

    @Mock
    ConnectionProvider connectionProvider;
    @Mock
    BasePresenterDependency basePresenterDependency;

    MockPresenter presenter;

    private Single<String> single = Single.just("Single");
    private Maybe<String> maybe = Maybe.just("Maybe");
    private Maybe maybeAction = Maybe.fromAction(() -> {
        int i = 1;
    });
    private Completable completable = Completable.complete();

    private Throwable testThrowable = new Throwable("test");
    private Single<String> singleError = Single.error(testThrowable);
    private Maybe<String> maybeError = Maybe.error(testThrowable);
    private Completable completableError = Completable.error(testThrowable);

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);
        PowerMockito.mockStatic(Logger.class);

        SchedulersProvider schedulersProvider = new SchedulersProvider() {
            @Override
            public Scheduler main() {
                return Schedulers.trampoline(); //аналог immediate() в RxJava 2
            }

            @Override
            public Scheduler worker() {
                return Schedulers.trampoline(); //аналог immediate() в RxJava 2
            }
        };

        basePresenterDependency = new BasePresenterDependency(schedulersProvider,
                mock(ActivityScreenState.class),
                mock(BaseScreenEventDelegateManager.class),
                mock(ErrorHandler.class),
                connectionProvider,
                mock(ActivityNavigator.class));
        presenter = new MockPresenter(basePresenterDependency);
    }

    //region subscribeIoTests
    @Test
    public void subscribeSingle() {

        ConsumerSafe<String> onSuccess = mock(ConsumerSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);

        presenter.subscribeIo(single, onSuccess, onError);
        verify(onSuccess, only()).accept("Single");
    }

    @Test
    public void subscribeCompletable() {
        ActionSafe onComplete = mock(ActionSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);

        presenter.subscribeIo(completable, onComplete, onError);
        verify(onComplete, only()).run();
    }

    @Test
    public void subscribeMaybe() {
        ConsumerSafe<String> onSuccess = mock(ConsumerSafe.class);
        ActionSafe onComplete = mock(ActionSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);

        presenter.subscribeIo(maybe, onSuccess, onComplete, onError);
        verify(onSuccess, only()).accept("Maybe");
    }

    @Test
    public void subscribeMaybeAction() {
        ConsumerSafe<String> onSuccess = mock(ConsumerSafe.class);
        ActionSafe onComplete = mock(ActionSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);

        presenter.subscribeIo(maybeAction, onSuccess, onComplete, onError);
        verify(onComplete, only()).run();
    }

    @Test
    public void subscribeSingleError() {

        ConsumerSafe<String> onSuccess = mock(ConsumerSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);

        presenter.subscribeIo(singleError, onSuccess, onError);
        verify(onError, only()).accept(testThrowable);
    }

    @Test
    public void subscribeCompletableError() {
        ActionSafe onComplete = mock(ActionSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);

        presenter.subscribeIo(completableError, onComplete, onError);
        verify(onError, only()).accept(testThrowable);
    }

    @Test
    public void subscribeMaybeError() {
        ConsumerSafe<String> onSuccess = mock(ConsumerSafe.class);
        ActionSafe onComplete = mock(ActionSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);

        presenter.subscribeIo(maybeError, onSuccess, onComplete, onError);
        verify(onError, only()).accept(testThrowable);
    }
    //endregion

    //region subscribeIoHandlerError
    @Test
    public void subscribeIoHandleErrorSingle() {

        ConsumerSafe<String> onSuccess = mock(ConsumerSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);

        presenter.subscribeIoHandleError(single, onSuccess, onError);
        verify(onSuccess, only()).accept("Single");
    }

    @Test
    public void subscribeIoHandleErrorCompletable() {
        ActionSafe onComplete = mock(ActionSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);

        presenter.subscribeIoHandleError(completable, onComplete, onError);
        verify(onComplete, only()).run();
    }

    @Test
    public void subscribeIoHandleErrorMaybe() {
        ConsumerSafe<String> onSuccess = mock(ConsumerSafe.class);
        ActionSafe onComplete = mock(ActionSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);

        presenter.subscribeIoHandleError(maybe, onSuccess, onComplete, onError);
        verify(onSuccess, only()).accept("Maybe");
    }

    @Test
    public void subscribeIoHandleErrorMaybeAction() {
        ConsumerSafe<String> onSuccess = mock(ConsumerSafe.class);
        ActionSafe onComplete = mock(ActionSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);

        presenter.subscribeIoHandleError(maybeAction, onSuccess, onComplete, onError);
        verify(onComplete, only()).run();
    }

    @Test
    public void subscribeIoHandleErrorSingleError() {

        ConsumerSafe<String> onSuccess = mock(ConsumerSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);

        presenter.subscribeIoHandleError(singleError, onSuccess, onError);
        verify(onError, only()).accept(testThrowable);
    }

    @Test
    public void subscribeIoHandleErrorCompletableError() {
        ActionSafe onComplete = mock(ActionSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);

        presenter.subscribeIo(completableError, onComplete, onError);
        verify(onError, only()).accept(testThrowable);
    }

    @Test
    public void subscribeIoHandleErrorMaybeError() {
        ConsumerSafe<String> onSuccess = mock(ConsumerSafe.class);
        ActionSafe onComplete = mock(ActionSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);

        presenter.subscribeIoHandleError(maybeError, onSuccess, onComplete, onError);
        verify(onError, only()).accept(testThrowable);
    }
    //endregion

    //region subscribeAutoReload
    @Test
    public void subscribeIoAutoReloadSingle() {

        ConsumerSafe<String> onSuccess = mock(ConsumerSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);
        ActionSafe reloadAction = mock(ActionSafe.class);

        doReturn(true).when(connectionProvider).isDisconnected();
        doReturn(Observable.just(true)).when(connectionProvider).observeConnectionChanges();

        presenter.subscribeIoAutoReload(singleError, reloadAction, onSuccess, onError);
        verify(reloadAction, only()).run();
    }

    @Test
    public void subscribeAutoReloadCompletable() {
        ActionSafe onComplete = mock(ActionSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);
        ActionSafe reloadAction = mock(ActionSafe.class);

        doReturn(true).when(connectionProvider).isDisconnected();
        doReturn(Observable.just(true)).when(connectionProvider).observeConnectionChanges();

        presenter.subscribeIoAutoReload(completableError, reloadAction, onComplete, onError);
        verify(reloadAction, only()).run();
    }

    @Test
    public void subscribeAutoReloadMaybe() {
        ConsumerSafe<String> onSuccess = mock(ConsumerSafe.class);
        ActionSafe onComplete = mock(ActionSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);
        ActionSafe reloadAction = mock(ActionSafe.class);

        doReturn(true).when(connectionProvider).isDisconnected();
        doReturn(Observable.just(true)).when(connectionProvider).observeConnectionChanges();

        presenter.subscribeIoAutoReload(maybeError, reloadAction, onSuccess, onComplete, onError);
        verify(reloadAction, only()).run();
    }
    //endregion

    //region subscribeAutoReloadHandleError
    @Test
    public void subscribeIoHandleErrorAutoReloadSingle() {

        ConsumerSafe<String> onSuccess = mock(ConsumerSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);
        ActionSafe reloadAction = mock(ActionSafe.class);

        doReturn(true).when(connectionProvider).isDisconnected();
        doReturn(Observable.just(true)).when(connectionProvider).observeConnectionChanges();

        presenter.subscribeIoHandleErrorAutoReload(singleError, reloadAction, onSuccess, onError);
        verify(reloadAction, only()).run();
        verify(onError, only()).accept(testThrowable);
    }

    @Test
    public void subscribeHandleErrorAutoReloadCompletable() {
        ActionSafe onComplete = mock(ActionSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);
        ActionSafe reloadAction = mock(ActionSafe.class);

        doReturn(true).when(connectionProvider).isDisconnected();
        doReturn(Observable.just(true)).when(connectionProvider).observeConnectionChanges();

        presenter.subscribeIoHandleErrorAutoReload(completableError, reloadAction, onComplete, onError);
        verify(reloadAction, only()).run();
        verify(onError, only()).accept(testThrowable);
    }

    @Test
    public void subscribeHandleErrorAutoReloadMaybe() {
        ConsumerSafe<String> onSuccess = mock(ConsumerSafe.class);
        ActionSafe onComplete = mock(ActionSafe.class);
        ConsumerSafe<Throwable> onError = mock(ConsumerSafe.class);
        ActionSafe reloadAction = mock(ActionSafe.class);

        doReturn(true).when(connectionProvider).isDisconnected();
        doReturn(Observable.just(true)).when(connectionProvider).observeConnectionChanges();

        presenter.subscribeIoHandleErrorAutoReload(maybeError, reloadAction, onSuccess, onComplete, onError);
        verify(reloadAction, only()).run();
        verify(onError, only()).accept(testThrowable);
    }
    //endregion
}

File: core-mvp/src/test/java/MockPresenterTest.kt
import io.reactivex.*
import io.reactivex.schedulers.Schedulers
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.doReturn
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations
import org.mockito.internal.verification.VerificationModeFactory.only
import org.powermock.api.mockito.PowerMockito
import org.powermock.api.mockito.PowerMockito.mock
import org.powermock.core.classloader.annotations.PrepareForTest
import org.powermock.modules.junit4.PowerMockRunner
import ru.surfstudio.android.connection.ConnectionProvider
import ru.surfstudio.android.core.mvp.error.ErrorHandler
import ru.surfstudio.android.core.mvp.presenter.BasePresenter
import ru.surfstudio.android.core.mvp.presenter.BasePresenterDependency
import ru.surfstudio.android.core.ui.event.BaseScreenEventDelegateManager
import ru.surfstudio.android.core.ui.navigation.activity.navigator.ActivityNavigator
import ru.surfstudio.android.core.ui.state.ActivityScreenState
import ru.surfstudio.android.logger.Logger
import ru.surfstudio.android.rx.extension.scheduler.SchedulersProvider

/**
 * unit tests for [BasePresenter]
 */

@RunWith(PowerMockRunner::class)
@PrepareForTest(Logger::class)
class MockPresenterTest {

    @Mock
    var connectionProvider: ConnectionProvider? = null
    @Mock
    lateinit var basePresenterDependency: BasePresenterDependency
    @Mock
    private val onSuccess: (String) -> Unit = {}
    @Mock
    private val onSuccessMayBe: (Any) -> Unit = { }
    @Mock
    private val onError: (Throwable) -> Unit = {}
    @Mock
    private val onComplete: () -> Unit = {}
    @Mock
    private val reloadAction: () -> Unit = {}

    internal lateinit var presenter: MockPresenter

    private val single = Single.just("Single")
    private val maybe = Maybe.just("Maybe")
    private val maybeAction = Maybe.fromAction<Any> { val i = 1 }
    private val completable = Completable.complete()

    private val testThrowable = Throwable("test")
    private val singleError = Single.error<String>(testThrowable)
    private val maybeError = Maybe.error<String>(testThrowable)
    private val completableError = Completable.error(testThrowable)

    @Before
    @Throws(Exception::class)
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        PowerMockito.mockStatic(Logger::class.java)

        val schedulersProvider = object : SchedulersProvider {
            override fun main(): Scheduler {
                return Schedulers.trampoline() //аналог immediate() в RxJava 2
            }

            override fun worker(): Scheduler {
                return Schedulers.trampoline() //аналог immediate() в RxJava 2
            }
        }

        basePresenterDependency = BasePresenterDependency(schedulersProvider,
                mock(ActivityScreenState::class.java),
                mock(BaseScreenEventDelegateManager::class.java),
                mock(ErrorHandler::class.java),
                connectionProvider,
                mock(ActivityNavigator::class.java))
        presenter = MockPresenter(basePresenterDependency)
    }

    //region subscribeIoTests
    @Test
    fun subscribeSingle() {
        presenter.testSubscribeIo(single, onSuccess, onError)
        verify(onSuccess, only()).invoke("Single")
    }

    @Test
    fun subscribeCompletable() {
        presenter.testSubscribeIo(completable, onComplete, onError)
        verify(onComplete, only()).invoke()
    }

    @Test
    fun subscribeMaybe() {
        presenter.testSubscribeIo(maybe, onSuccess, onComplete, onError)
        verify<(String) -> Unit>(onSuccess, only()).invoke("Maybe")
    }

    @Test
    fun subscribeMaybeAction() {
        val onSuccess: (Any) -> Unit = { }

        presenter.testSubscribeIo(maybeAction, onSuccess, onComplete, onError)
        verify(onComplete, only()).invoke()
    }

    @Test
    fun subscribeSingleError() {
        presenter.testSubscribeIo(singleError, onSuccess, onError)
        verify<(Throwable) -> Unit>(onError, only()).invoke(testThrowable)
    }

    @Test
    fun subscribeCompletableError() {
        presenter.testSubscribeIo(completableError, onComplete, onError)
        verify<(Throwable) -> Unit>(onError, only()).invoke(testThrowable)
    }

    @Test
    fun subscribeMaybeError() {
        val onSuccess: (Any) -> Unit = { }

        presenter.testSubscribeIo(maybeError, onSuccess, onComplete, onError)
        verify<(Throwable) -> Unit>(onError, only()).invoke(testThrowable)
    }
    //endregion

    //region subscribeIoHandlerError
    @Test
    fun subscribeIoHandleErrorSingle() {
        presenter.testSubscribeIoHandleError(single, onSuccess, onError)
        verify<(String) -> Unit>(onSuccess, only()).invoke("Single")
    }

    @Test
    fun subscribeIoHandleErrorCompletable() {
        presenter.testSubscribeIoHandleError(completable, onComplete, onError)
        verify(onComplete, only()).invoke()
    }

    @Test
    fun subscribeIoHandleErrorMaybe() {
        presenter.testSubscribeIoHandleError(maybe, onSuccessMayBe, onComplete, onError)
        verify<(String) -> Unit>(onSuccessMayBe, only()).invoke("Maybe")
    }

    @Test
    fun subscribeIoHandleErrorMaybeAction() {
        presenter.testSubscribeIoHandleError(maybeAction, onSuccessMayBe, onComplete, onError)
        verify(onComplete, only()).invoke()
    }

    @Test
    fun subscribeIoHandleErrorSingleError() {
        presenter.testSubscribeIoHandleError(singleError, onSuccess, onError)
        verify<(Throwable) -> Unit>(onError, only()).invoke(testThrowable)
    }

    @Test
    fun subscribeIoHandleErrorCompletableError() {
        presenter.testSubscribeIo(completableError, onComplete, onError)
        verify<(Throwable) -> Unit>(onError, only()).invoke(testThrowable)
    }

    @Test
    fun subscribeIoHandleErrorMaybeError() {
        presenter.testSubscribeIoHandleError(maybeError, onSuccessMayBe, onComplete, onError)
        verify<(Throwable) -> Unit>(onError, only()).invoke(testThrowable)
    }
    //endregion

    //region subscribeAutoReload
    @Test
    fun subscribeIoAutoReloadSingle() {
        doReturn(true).`when`<ConnectionProvider>(connectionProvider).isDisconnected
        doReturn(Observable.just(true)).`when`<ConnectionProvider>(connectionProvider).observeConnectionChanges()

        presenter.testSubscribeIoAutoReload(singleError, reloadAction, onSuccess, onError)
        verify(reloadAction, only()).invoke()
    }

    @Test
    fun subscribeAutoReloadCompletable() {
        doReturn(true).`when`<ConnectionProvider>(connectionProvider).isDisconnected
        doReturn(Observable.just(true)).`when`<ConnectionProvider>(connectionProvider).observeConnectionChanges()

        presenter.testSubscribeIoAutoReload(completableError, reloadAction, onComplete, onError)
        verify(reloadAction, only()).invoke()
    }

    @Test
    fun subscribeAutoReloadMaybe() {
        doReturn(true).`when`<ConnectionProvider>(connectionProvider).isDisconnected
        doReturn(Observable.just(true)).`when`<ConnectionProvider>(connectionProvider).observeConnectionChanges()

        presenter.testSubscribeIoAutoReload(maybeError, reloadAction, onSuccess, onComplete, onError)
        verify(reloadAction, only()).invoke()
    }
    //endregion

    //region subscribeAutoReloadHandleError
    @Test
    fun subscribeIoHandleErrorAutoReloadSingle() {
        doReturn(true).`when`<ConnectionProvider>(connectionProvider).isDisconnected
        doReturn(Observable.just(true)).`when`<ConnectionProvider>(connectionProvider).observeConnectionChanges()

        presenter.testSubscribeIoHandleErrorAutoReload(singleError, reloadAction, onSuccess, onError)
        verify(reloadAction, only()).invoke()
        verify<(Throwable) -> Unit>(onError, only()).invoke(testThrowable)
    }

    @Test
    fun subscribeHandleErrorAutoReloadCompletable() {
        doReturn(true).`when`<ConnectionProvider>(connectionProvider).isDisconnected
        doReturn(Observable.just(true)).`when`<ConnectionProvider>(connectionProvider).observeConnectionChanges()

        presenter.testSubscribeIoHandleErrorAutoReload(completableError, reloadAction, onComplete, onError)
        verify(reloadAction, only()).invoke()
        verify<(Throwable) -> Unit>(onError, only()).invoke(testThrowable)
    }

    @Test
    fun subscribeHandleErrorAutoReloadMaybe() {
        doReturn(true).`when`<ConnectionProvider>(connectionProvider).isDisconnected
        doReturn(Observable.just(true)).`when`<ConnectionProvider>(connectionProvider).observeConnectionChanges()

        presenter.testSubscribeIoHandleErrorAutoReload(maybeError, reloadAction, onSuccess, onComplete, onError)
        verify(reloadAction, only()).invoke()
        verify<(Throwable) -> Unit>(onError, only()).invoke(testThrowable)
    }
    //endregion
}
